clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreBufferService.java

```diff
@@ -472,138 +201,421 @@ public void drainBufferedRecordsFromTopicPartition(PubSubTopicPartition topicPar
+    String errorMessage = "There are still some records left in the blocking queue of store writer thread: "
+        + workerIndex + " for topic: " + topicPartition.getPubSubTopic().getName() + " partition after retry for "
+        + retryNum + " times";
+    LOGGER.error(errorMessage);
+    throw new VeniceException(errorMessage);
+  }
+
+  @Override
+  public boolean startInner() {
+    this.executorService = Executors.newFixedThreadPool(drainerNum, new DaemonThreadFactory("Store-writer"));
+
+    // Submit all the buffer drainers
+    for (int cur = 0; cur < drainerNum; ++cur) {
+      StoreBufferDrainer drainer = new StoreBufferDrainer(this.blockingQueueArr.get(cur), cur, storeBufferServiceStats);
+      this.executorService.submit(drainer);
+      drainerList.add(drainer);
+    }
+    this.executorService.shutdown();
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    // Graceful shutdown
+    drainerList.forEach(drainer -> drainer.stop());
+    if (this.executorService != null) {
+      this.executorService.shutdownNow();
+      this.executorService.awaitTermination(10, TimeUnit.SECONDS);
+    }
+  }
+
+  @Override
+  public int getDrainerCount() {
+    return blockingQueueArr.size();
+  }
+
+  @Override
+  public long getDrainerQueueMemoryUsage(int index) {
+    return blockingQueueArr.get(index).getMemoryUsage();
+  }
+
+  @Override
+  public long getTotalMemoryUsage() {
+    long totalUsage = 0;
+    for (MemoryBoundBlockingQueue<QueueNode> queue: blockingQueueArr) {
+      totalUsage += queue.getMemoryUsage();
+    }
+    return totalUsage;
+  }
+
+  @Override
+  public long getTotalRemainingMemory() {
+    long totalRemaining = 0;
+    for (MemoryBoundBlockingQueue<QueueNode> queue: blockingQueueArr) {
+      totalRemaining += queue.remainingMemoryCapacityInByte();
+    }
+    return totalRemaining;
+  }
+
+  @Override
+  public long getMaxMemoryUsagePerDrainer() {
+    long maxUsage = 0;
+    boolean slowDrainerExists = false;
+
+    for (MemoryBoundBlockingQueue<QueueNode> queue: blockingQueueArr) {
+      maxUsage = Math.max(maxUsage, queue.getMemoryUsage());
+      if (queue.getMemoryUsage() > 0.8 * bufferCapacityPerDrainer) {
+        slowDrainerExists = true;
+      }
+    }
+
+    for (int index = 0; index < blockingQueueArr.size(); index++) {
+      StoreBufferDrainer drainer = drainerList.get(index);
+      // print drainer info when there is a slow drainer.
+      if (slowDrainerExists) {
+        MemoryBoundBlockingQueue<QueueNode> queue = blockingQueueArr.get(index);
+        int count = queue.getMemoryUsage() > 0.8 * bufferCapacityPerDrainer ? 5 : 1;
+        List<Map.Entry<PubSubTopicPartition, Long>> slowestEntries = drainer.topicToTimeSpent.entrySet()
+            .stream()
+            .sorted(comparing(Map.Entry::getValue, reverseOrder()))
+            .limit(count)
+            .collect(toList());
+        // TODO: break this done so we can let it emit latency metrics too.
```

Comment:
NIT: typo in the comment but also wondering if the existence of a slow drainer itself is worth a metric.

Reply:
This is a old comment since I was moving some codes around... Let me remove this.
To your questions, no I don't think we need to add a specific metric here as we are already adding some processing latency metrics at the store-level and having them at the drainer-level is a duplicate effort.